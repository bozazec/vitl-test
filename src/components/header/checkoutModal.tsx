import React from 'react'
import {Modal} from "@fluentui/react";
import {contentModalStyles} from "../addToCartModal/addToCartModal";

interface ICheckoutProps {
	isOpen: boolean
}
const CheckoutModal = (props: ICheckoutProps) => {
	const {isOpen} = props
	return <Modal
		isOpen={isOpen}
		containerClassName={contentModalStyles.container}
	>
		<div className={contentModalStyles.header}>
			<div id={'title'}>Fantastic</div>
		</div>
		<div className={contentModalStyles.body}>
			<div className={'subtitle'}>Your order is on the way</div>
		</div>
	</Modal>
}

export default CheckoutModal