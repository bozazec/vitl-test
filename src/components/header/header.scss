.header {
  position: fixed;
  text-align: center;
  top: 0;
  width: 100%;
  height: 40px;
  color: #ffffff;
  z-index: 1;

  .logo {
    margin-top: 4px;
    height: 25px;
  }

  .update {
    position: absolute;
    right: 60px;
    top: 8px;
    font-size: 20px;
    cursor: pointer;
    &:hover {
      opacity: 0.9;
    }
    &:active {
      opacity: 0.7;
    }
  }

  .cart {
    position: absolute;
    right: 20px;
    top: 8px;
    font-size: 20px;
    cursor: pointer;
    &:hover {
      opacity: 0.9;
    }
    &:active {
      opacity: 0.7;
    }
  }

  .badge {
    position: absolute;
    top: 3px;
    right: 10px;
    width: 10px;
    background: #ffd326;
    color: black;
    font-size: 12px;
    border-radius: 50%;
    padding: 0 4px;
  }


  .cartContainer {
    padding: 5px 10px;
    transition: all 0.5ms ease-in-out;
    position: absolute;
    right: -450px;
    top: 40px;
    color: black;
    width: 350px;
    overflow: hidden;
    background: #ffd326;
    font-size: 12px;

    &.open {
      right: 0;
    }


    .cartHeader {
      text-align: left;
      display: flex;
      font-weight: 600;
      margin-bottom: 10px;
    }

    .cartItem {
      text-align: left;
      display: flex;
    }

    .name {
      min-width: 170px;

    }

    .qty {
      min-width: 60px;
      display: flex;
      justify-content: space-between;
      margin-right: 25px;
      align-items: center;

      i {
        cursor: pointer;

        &:hover {
          opacity: 0.8;
        }

        &:active {
          opacity: 1
        }
      }
    }

    .price {
      min-width: 50px;
    }

    .total {
      margin-top: 15px;
      display: flex;
      justify-content: flex-end;
      font-size: 14px;
      font-weight: 600;

      .title {
        margin-right: 15px;
      }
    }

    button {
      width: 370px;
      margin: 15px -15px -5px -15px;
    }
  }
   @media only screen and (max-width: 500px) {
      .cartContainer {
        width: 300px;
      }
      button {
        width: 320px;
      }
    }
}
