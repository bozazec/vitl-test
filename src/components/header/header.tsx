import React, {useEffect, useState} from 'react'
import logo from "../../assets/images/logo.svg";
import {Icon, PrimaryButton, useTheme} from "@fluentui/react";
import {useDispatch, useSelector} from "react-redux";
import {IAppState} from "../../store";
import {HomeActionTypes} from "../../pages/home/homeReducer";
import CheckoutModal from "./checkoutModal";
import './header.scss'
import CartItem from "../cartItem/cartItem";


const Header = () => {
	const theme = useTheme()
	const dispatch = useDispatch()
	const cart = useSelector((state: IAppState) => state.home.cart)
	const config = useSelector((state: IAppState) => state.home.config)
	const update = useSelector((state: IAppState) => state.home.updateAvailable)
	const [qty, setQty] = useState(0)
	const [showCart, setShowCart] = useState(false)
	const [total, setTotal] = useState(0)

	const [showCheckoutModal, setShowCheckoutModal] = useState(false)

	useEffect(() => {
		let quantity = 0
		let _total = 0
		Object.keys(cart).forEach((key: string) => {
			quantity = quantity + cart[key].quantity
			_total = _total + (cart[key].price * cart[key].quantity)
		})
		setQty(quantity)
		setTotal(_total)
	}, [cart])

	const checkout = () => {
		dispatch({
			type: HomeActionTypes.SET_CART,
			payload: {}
		})
		setShowCheckoutModal(true)
		setTimeout(() => {
			setShowCheckoutModal(false)
		}, 2000)
	}

	// Update the service worker and reload the app
	const updateApp = () => {
		const registrationWaiting = update?.waiting;
		if (registrationWaiting) {
			registrationWaiting.postMessage({type: 'SKIP_WAITING'});
			registrationWaiting.addEventListener('statechange', (e: any) => {
				if (e.target.state === 'activated') {
					window.location.reload();
				}
			});
		}
	}

	return <header className="header" style={{backgroundColor: theme.palette.themePrimary}}>
		<img className={'logo'} src={logo} alt={'vitl'}/>
		{/*Update button*/}
		{update && <Icon iconName={'CloudDownload'} className={'update'} onClick={updateApp}/>}
		{/*Cart button*/}
		<Icon onClick={() => setShowCart(!showCart)} className={'cart'} iconName={'ShoppingCart'} ariaLabel={'cart'}/>

		{/*Badge*/}
		{Object.keys(cart).length > 0 ? <div className={'badge'} onClick={() => setShowCart(!showCart)}>{qty}</div> : null}


		{Object.keys(cart).length > 0 ? <div>
			{/*Cart header*/}
			<div className={`cartContainer ${showCart ? 'open' : ''}`} style={{height: 'fit-content'}}>
				<div className={'cartHeader'}>
					<div className={'name'}>Item</div>
					<div className={'qty'}>Qty</div>
					<div className={'price'}>$</div>
				</div>

				{/*Cart items*/}
				{Object.values(cart).map((item: any) => <CartItem key={item.name} {...item} config={config}/>)}

				{/*Total amount to pay*/}
				<div className={'total'}>
					<div className={'title'}>Total</div>
					<div>{total.toFixed(1)} {config.currency}</div>
				</div>

				{/*Checkout button*/}
				<PrimaryButton text={'Checkout'} onClick={checkout}/>
			</div>
		</div> : null}
		<CheckoutModal isOpen={showCheckoutModal}/>
	</header>
}

export default Header

