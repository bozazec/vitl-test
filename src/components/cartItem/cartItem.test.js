import {render, screen, fireEvent, waitFor, act} from "@testing-library/react";
import {Provider} from "react-redux";
import React from "react";
import CartItem from "./cartItem";
import {mockReduxStore} from "../header/header.test";


const mockConfig = {
	"tolerableUpperLimits": [
		{
			"id": "vitamin-a",
			"amount": 1500,
			"unit": "mcg"
		},
		{
			"id": "vitamin-c",
			"amount": 1000,
			"unit": "mg"
		},
		{
			"id": "vitamin-d",
			"amount": 75,
			"unit": "mcg"
		},
		{
			"id": "vitamin-e",
			"amount": 540,
			"unit": "mg"
		},
		{
			"id": "zinc",
			"amount": 25,
			"unit": "mg"
		}
	],
	"currency": "GBP"
}

const mockItemProps = {
	name: 'Vitamin D',
	price: 10,
	quantity: 1,
	config: mockConfig
}

describe('CartItem', () => {
	// Test using render and screen from @testing-library/react
	test('Rendering', () => {
		const {container} = render(<Provider store={mockReduxStore}>
			<CartItem {...mockItemProps}/>
		</Provider>)

		expect(container).toMatchSnapshot()
	})

	test('Add item', async () => {
		jest.useFakeTimers()
		render(<Provider store={mockReduxStore}>
			<CartItem {...mockItemProps}/>
		</Provider>)

		const addItemButton = screen.getByTestId('addItemButton')
		fireEvent.click(addItemButton)
		await act(async () => {
				jest.runOnlyPendingTimers();
			});

		setTimeout(async () => {
			const qty = await screen.findByTestId('quantityTestId')

			expect(qty).toHaveTextContent('2')
		}, 1000)

	})
})