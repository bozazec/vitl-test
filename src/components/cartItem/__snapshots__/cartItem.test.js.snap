// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem Rendering 1`] = `
<div>
  <div
    class="cartItem"
  >
    <div
      class="name"
    >
      Vitamin D
    </div>
    <div
      class="qty"
      data-testid="qty"
    >
      <i
        aria-hidden="true"
        class="root-118"
        data-icon-name="Remove"
      >
        
      </i>
      <span
        data-testid="quantityTestId"
      >
        1
      </span>
      <i
        aria-hidden="true"
        class="root-121"
        data-icon-name="Add"
        data-testid="addItemButton"
      >
        
      </i>
    </div>
    <div
      class="price"
    >
      10.0
       
      GBP
    </div>
  </div>
</div>
`;

exports[`Header Snapshot with right 450 1`] = `
<div>
  <header
    class="header"
    style="background-color: rgb(0, 120, 212);"
  >
    <img
      alt="vitl"
      class="logo"
      data-testid="logo"
      src="logo.svg"
    />
    <i
      aria-label="cart"
      class="cart root-115"
      data-icon-name="ShoppingCart"
      data-testid="cartButton"
      role="img"
    >
      
    </i>
    <div
      class="badge"
    >
      3
    </div>
    <div>
      <div
        class="cartContainer "
        data-testid="cartContainer"
      >
        <div
          class="cartHeader"
        >
          <div
            class="name"
          >
            Item
          </div>
          <div
            class="qty"
          >
            Qty
          </div>
          <div
            class="price"
          >
            Cost
          </div>
        </div>
        <div
          class="cartItem"
        >
          <div
            class="name"
          >
            Vitamin D
          </div>
          <div
            class="qty"
            data-testid="qty"
          >
            <i
              aria-hidden="true"
              class="root-118"
              data-icon-name="Remove"
            >
              
            </i>
            <span
              data-testid="quantityTestId"
            >
              1
            </span>
            <i
              aria-hidden="true"
              class="root-121"
              data-icon-name="Add"
              data-testid="addItemButton"
            >
              
            </i>
          </div>
          <div
            class="price"
          >
            2.3
             
            GBP
          </div>
        </div>
        <div
          class="cartItem"
        >
          <div
            class="name"
          >
            Vitamin C
          </div>
          <div
            class="qty"
            data-testid="qty"
          >
            <i
              aria-hidden="true"
              class="root-118"
              data-icon-name="Remove"
            >
              
            </i>
            <span
              data-testid="quantityTestId"
            >
              1
            </span>
            <i
              aria-hidden="true"
              class="root-121"
              data-icon-name="Add"
              data-testid="addItemButton"
            >
              
            </i>
          </div>
          <div
            class="price"
          >
            4.5
             
            GBP
          </div>
        </div>
        <div
          class="cartItem"
        >
          <div
            class="name"
          >
            Vitamin A
          </div>
          <div
            class="qty"
            data-testid="qty"
          >
            <i
              aria-hidden="true"
              class="root-118"
              data-icon-name="Remove"
            >
              
            </i>
            <span
              data-testid="quantityTestId"
            >
              1
            </span>
            <i
              aria-hidden="true"
              class="root-121"
              data-icon-name="Add"
              data-testid="addItemButton"
            >
              
            </i>
          </div>
          <div
            class="price"
          >
            6.0
             
            GBP
          </div>
        </div>
        <div
          class="total"
        >
          <div
            class="title"
          >
            Total
          </div>
          <div>
            12.8
             
            GBP
          </div>
        </div>
        <button
          class="ms-Button ms-Button--primary root-122"
          data-is-focusable="true"
          type="button"
        >
          <span
            class="ms-Button-flexContainer flexContainer-123"
            data-automationid="splitbuttonprimary"
          >
            <span
              class="ms-Button-textContainer textContainer-124"
            >
              <span
                class="ms-Button-label label-126"
                id="id__4"
              >
                Checkout
              </span>
            </span>
          </span>
        </button>
      </div>
    </div>
  </header>
</div>
`;
